pretyping/tacred.ml

```diff
@@ -761,30 +767,29 @@ and whd_simpl_stack env sigma =
 	  with Redelimination -> s')
 
       | Proj (p, c) ->
-        (try 
-	   let unf = Projection.unfolded p in
-	     if unf || is_evaluable env (EvalConstRef (Projection.constant p)) then
-               let npars = Projection.npars p in
- 		 (match unf, ReductionBehaviour.get (ConstRef (Projection.constant p)) with
- 		 | false, Some (l, n, f) when List.mem `ReductionNeverUnfold f -> 
-                   (* simpl never *) s'
-		 | false, Some (l, n, f) when not (List.is_empty l) ->
-		   let l' = List.map_filter (fun i -> 
-                     let idx = (i - (npars + 1)) in
-		       if idx < 0 then None else Some idx) l in
-		   let stack = reduce_params env sigma stack l' in
-                     (match reduce_projection env sigma p ~npars
-		       (whd_construct_stack env sigma c) stack 
-		      with
-		      | Reduced s' -> redrec (applist s')
-		      | NotReducible -> s')
- 		 | _ ->
-                   match reduce_projection env sigma p ~npars (whd_construct_stack env sigma c) stack with
-		   | Reduced s' -> redrec (applist s')
-		   | NotReducible -> s')
-	   else s'
-	 with Redelimination -> s')
-	  
+        (try
+           let unf = Projection.unfolded p in
+           if unf || is_evaluable env (EvalConstRef (Projection.constant p)) then
+             let npars = Projection.npars p in
+             (match unf, ReductionBehaviour.get (ConstRef (Projection.constant p)) with
+              | false, Some ReductionBehaviour.NeverUnfold -> s'
+              | false, Some (ReductionBehaviour.UnfoldWhen { recargs }) when not (List.is_empty recargs) -> (* TODO why are nargs and dontexposecase ignored? *)
```

Comment:
@gares do you know why only recargs is taken into account here?

Reply:
No, when I added the reduction behavior PrimProj was not there.. These lines come from 84544396cbbf34848be2240acf181b4d5f1f42d2 .